File: lib/drawing-config.js
"use strict";
// Single source of truth for the entire drawing model.
// Server requires() this directly (zero latency). Clients fetch /api/drawing-config.
// DB key "drawing_config" stores a JSON object whose fields OVERRIDE these defaults.
// Change any default here → redeploy. Change any value at runtime → POST /api/admin/drawing-config.

const DEFAULTS = {

  // ── 1. DIFFICULTY TIERS ───────────────────────────────────────────────────────
  // minArea:      minimum pixel count for a segmented region to be kept (not discarded as noise).
  //               Smaller = more regions survive = busier image. Geometric midpoint of prior web/Flutter values.
  // maxRegions:   cap on how many numbered regions are shown. Excess regions become free-colour only.
  // colorCount:   default palette slots displayed for this difficulty.
  // resolution:   generation width = height in pixels. Both platforms use this.
  // skipFreeTiers: true → bypass Pollinations/Together/Cloudflare, go straight to Novita.
  //               Hard/Extreme skip free tiers because free cold-misses each burn 45-50s and
  //               the 80s client budget runs out before Novita even starts.
  difficulties: {
    easy:    { minArea: 2000, maxRegions: 10, colorCount:  6, resolution: 1024, skipFreeTiers: true },
    medium:  { minArea:  800, maxRegions: 18, colorCount:  8, resolution: 1024, skipFreeTiers: true },
    hard:    { minArea:  300, maxRegions: 30, colorCount: 12, resolution: 1024, skipFreeTiers: true  },
    extreme: { minArea:   80, maxRegions: 48, colorCount: 99, resolution: 1024, skipFreeTiers: true  },
  },

  // ── 2. SIZE OPTIONS ───────────────────────────────────────────────────────────
  // w/h: pixel dimensions sent to the generation API (must be 512, 768, or 1024).
  // promptHint: appended to the base prompt so the model knows scene extent. null = no hint.
  // Note: large and xxl share dimensions — they differ only in promptHint (more background detail).
  sizes: {
    small:  { w:  512, h:  512, promptHint: "single centered subject, sparse simple layout" },
    medium: { w:  768, h:  768, promptHint: null },
    large:  { w: 1024, h: 1024, promptHint: "wide scene filling the full canvas, include background elements" },
    xxl:    { w: 1024, h: 1024, promptHint: "panoramic wide scene filling the entire canvas, rich background detail throughout" },
  },
  defaultSize: "xxl",    // web default (user-selectable)
  mobileSize:  "large",  // Flutter default — 1024×1024 (was hardcoded 768×768)

  // ── 3. REGION DETECTION ───────────────────────────────────────────────────────
  detection: {

    // Outline mask — adaptive hysteresis thresholds for pixel classification.
    // All brightness values are 0 (black) to 255 (white).
    // globalDark:   brightness < this = definite black line (always a wall).
    // localMargin:  strong-line zone: adaptive_mean - localMargin. Catches anti-aliased edges.
    // adaptiveCeil: cap on the strong-line classifier (web=150, Flutter=190 → 165).
    //               Pixels brighter than this are never a strong line regardless of local mean.
    // weakMargin:   weak-candidate zone: adaptive_mean - weakMargin.
    // weakCeil:     near-white free threshold (web=205, Flutter=220 → 212).
    //               Pixels brighter than this are definitively free (colorable), not walls.
    outlineMask: {
      globalDark:   100,
      localMargin:   22,
      adaptiveCeil: 165,
      weakMargin:     8,
      weakCeil:     212,
    },

    // Line gap bridging — connects broken outline segments so regions are fully enclosed.
    // maxGap = clamp(imageShortEdge / 100, maxGapMin, maxGapMax) pixels.
    // faceCos: line-end endpoints must face each other within cos(60°) = 0.5 to be bridged.
    //          Prevents bridging parallel lines that run next to each other.
    // tangentSteps: how many steps along the tangent to probe before bridging.
    // maxGapMax: web=16px, Flutter=24px → 20px. Flutter allowed longer bridges to stitch
    //            JPEG-compressed seams; 20px is a principled midpoint.
    lineBridge: {
      maxGapMin:    4,
      maxGapMax:   20,
      faceCos:    0.5,
      tangentSteps: 4,
    },

    // Trapped-ball segmentation — ball radii series for region seeding.
    // maxR = clamp(imageShortEdge / 200, maxRMin, maxRMax).
    // At 1024px: maxR = clamp(5, 2, 7) = 5 → series 5→3→2→1.
    // decayNumerator/Denominator: next radius = floor(r * 2/3).
    trappedBall: {
      maxRMin:          2,
      maxRMax:          7,
      decayNumerator:   2,
      decayDenominator: 3,
      finalPassRadius:  1,
    },

    // Region validity filters applied after segmentation.
    // absoluteFloor: discard any region with fewer pixels than this. Web had 30 hardcoded;
    //                now explicit and tunable. Prevents stray 1-pixel dots from becoming tap targets.
    // promoteFloor:  background-adjacent specks above this size are promoted to real regions
    //                rather than merged into the background. Flutter had 50; now applied to both.
    // rescueFloor:   if ZERO valid regions survive (e.g., extreme intricate image where every
    //                trapped-ball seed is tiny), rescue the largest regions down to rescueFloor px
    //                so the image stays colorable rather than going blank.
    regionFilter: {
      absoluteFloor: 30,
      promoteFloor:  50,
      rescueFloor:    5,
    },

    // BFS fallback flood fill (pre-worker, before region segmentation completes).
    // bfsDarkThreshold: brightness < this = outline pixel = BFS wall. Web used 80 in the
    //                   fallback path vs 100 in the mask — kept at 80 for conservatism.
    // bfsFloodCapRatio: BFS may not flood more than this fraction of image pixels.
    //                   Prevents BFS from consuming the entire white background.
    bfsDarkThreshold: 80,
    bfsFloodCapRatio: 0.65,

    // workerTimeoutMs: how long the Web Worker / Flutter Isolate is allowed to run before
    //                  being killed and the canvas shown without region data.
    workerTimeoutMs: 90000,

    // snapRadius: tap-to-region lookup snaps to the nearest region within this pixel radius
    //             when the exact tap point lands on a line pixel.
    //             Platform-specific because canvas coordinate systems differ:
    //             web uses canvas-pixel space; Flutter uses logical px through fitImageRect.
    snapRadius: { web: 16, flutter: 12 },
  },

  // ── 4. COLOR ASSIGNMENT ───────────────────────────────────────────────────────
  // useGraphColoring:  Welsh-Powell adjacency-aware coloring so neighbouring regions
  //                    never share the same palette color. Web had this; Flutter did not → add to both.
  // useSemanticOrder:  reorder the palette so the subject's most characteristic color appears
  //                    first (e.g., green for frogs). Web had this; Flutter did not → add to both.
  // defaultColorCount: how many palette slots to show if the user hasn't changed the setting.
  // colorCountOptions: the selectable color count values shown in the UI.
  colorAssignment: {
    useGraphColoring:  true,
    useSemanticOrder:  true,
    defaultColorCount: 12,
    colorCountOptions: [6, 12, 18, 24, 99],
  },

  // ── 5. PALETTE PACKS ──────────────────────────────────────────────────────────
  // unlock: number of completed pictures required to unlock this pack (0 = always available).
  palettePacks: {
    classic: { unlock:  0 },
    pastel:  { unlock:  0 },
    nature:  { unlock:  0 },
    neon:    { unlock:  5 },
    candy:   { unlock: 15 },
    galaxy:  { unlock: 30 },
  },

  // ── 6. COMPLETION ─────────────────────────────────────────────────────────────
  // freeCoverageRatio:      free-colour mode: picture is done when ≥ this fraction of
  //                         the meaningful target regions are colored. 0.90 = 90%.
  // freehandCoverThreshold: a freehand (pencil/paint) region counts as colored once this
  //                         fraction of its interior pixels are painted. Low = forgiving
  //                         because children colour unevenly and leave white gaps.
  // tinyRegionAutoRatio:    a region smaller than this fraction of total image pixels is
  //                         auto-completed (treated as done without user interaction).
  //                         0.0005 = 0.05% of pixels ≈ a 23×23px dot at 1024×1024.
  //                         Flutter had this; web did not → now applied to both.
  completion: {
    freeCoverageRatio:      0.90,
    freehandCoverThreshold: 0.45,
    tinyRegionAutoRatio:    0.0005,
  },

  // ── 7. CANVAS ─────────────────────────────────────────────────────────────────
  // maxDisplayEdgePx: downscale source images whose longest edge exceeds this.
  // undoStackMax:     maximum undo history depth.
  canvas: {
    maxDisplayEdgePx: 2048,
    undoStackMax:       20,
  },

  // ── 8. GENERATION ─────────────────────────────────────────────────────────────
  // clientTimeoutMs: hard abort on the /api/generate-image fetch (web=75000, Flutter=90000 → 80000).
  // seedRange:       random seed is drawn from [0, seedRange).
  // maxImageBytes:   server-side cap on upstream image response size.
  generation: {
    clientTimeoutMs: 80000,
    seedRange:       2_000_000_000,
    maxImageBytes:   25 * 1024 * 1024,
  },

  // ── 9. PROVIDERS ──────────────────────────────────────────────────────────────
  // pollinationsTimeoutMs: per-request timeout for Pollinations (free tier 1).
  //                        Override with POLLINATIONS_TIMEOUT_MS env var.
  // fetchTimeoutMs:        per-request timeout for all other providers.
  //                        Override with FETCH_TIMEOUT_MS env var.
  // novitaDailyCap:        max paid Novita images per calendar day per server instance.
  //                        Override with NOVITA_DAILY_CAP env var.
  // rateLimitMax:          max /api/generate-image requests per IP per rateLimitWindowMs.
  // models/steps/guidance: provider-specific generation parameters.
  providers: {
    pollinationsTimeoutMs: 50_000,
    fetchTimeoutMs:        45_000,
    novitaDailyCap:        300,
    rateLimitMax:          15,
    rateLimitWindowMs:     3_600_000,
    models: {
      pollinations: "flux",
      together:     "black-forest-labs/FLUX.1-schnell-Free",
      cloudflare:   "@cf/black-forest-labs/flux-1-schnell",
      novita:       "flux-1-schnell",
      huggingface:  "black-forest-labs/FLUX.1-schnell",
    },
    steps: {
      together:    4,
      cloudflare:  8,
      novita:      4,
      huggingface: 8,
    },
    guidance: {
      huggingface: 3.5,
    },
  },
};

// ── Runtime config: DEFAULTS deep-merged with DB overrides ────────────────────

let _cache   = null;
let _cacheAt = 0;
const CACHE_TTL_MS = 60_000;

function deepMerge(target, source) {
  if (!source || typeof source !== "object" || Array.isArray(source)) return target;
  const out = Object.assign({}, target);
  for (const key of Object.keys(source)) {
    if (
      source[key] !== null &&
      typeof source[key] === "object" &&
      !Array.isArray(source[key]) &&
      typeof target[key] === "object" &&
      target[key] !== null
    ) {
      out[key] = deepMerge(target[key], source[key]);
    } else if (source[key] !== undefined) {
      out[key] = source[key];
    }
  }
  return out;
}

// Asynchronously loads DB overrides and merges them with DEFAULTS.
// Caches the result for CACHE_TTL_MS. Falls back to DEFAULTS if DB is unavailable.
async function getDrawingConfig() {
  if (_cache && Date.now() - _cacheAt < CACHE_TTL_MS) return _cache;
  try {
    const db = require("./db");
    const raw = await db.getConfig("drawing_config", "{}");
    const overrides = JSON.parse(raw || "{}");
    _cache   = deepMerge(DEFAULTS, overrides);
    _cacheAt = Date.now();
  } catch {
    _cache   = DEFAULTS;
    _cacheAt = Date.now();
  }
  return _cache;
}

// Synchronous read — returns last-loaded cache (or DEFAULTS before first async load).
// Safe to call from non-async contexts in image-providers.js.
function getDrawingConfigSync() {
  return _cache || DEFAULTS;
}

// Bust the cache (called by admin endpoint after a write).
function invalidateDrawingConfigCache() {
  _cacheAt = 0;
}

module.exports = { DEFAULTS, getDrawingConfig, getDrawingConfigSync, invalidateDrawingConfigCache, deepMerge };

File: flutter_app/lib/core/drawing_config.dart
// Remote drawing configuration served from /api/drawing-config.
//
// Constructed via DrawingConfig.fromJson with per-field fallbacks, so any
// missing or malformed key keeps the hard-coded DrawingConfig.defaults.
// Used by DrawingConfigService to expose a singleton via DrawingConfigService.instance.

class DifficultyConfig {
  final int minArea;
  final int maxRegions;
  final int colorCount;
  final int resolution;
  final bool skipFreeTiers;

  const DifficultyConfig({
    required this.minArea,
    required this.maxRegions,
    required this.colorCount,
    required this.resolution,
    required this.skipFreeTiers,
  });

  factory DifficultyConfig.fromJson(
      Map<String, dynamic> json, DifficultyConfig d) {
    return DifficultyConfig(
      minArea: (json['minArea'] as num?)?.toInt() ?? d.minArea,
      maxRegions: (json['maxRegions'] as num?)?.toInt() ?? d.maxRegions,
      colorCount: (json['colorCount'] as num?)?.toInt() ?? d.colorCount,
      resolution: (json['resolution'] as num?)?.toInt() ?? d.resolution,
      skipFreeTiers: (json['skipFreeTiers'] as bool?) ?? d.skipFreeTiers,
    );
  }
}

class OutlineMaskConfig {
  final int globalDark;
  final int localMargin;
  final int adaptiveCeil;
  final int weakMargin;
  final int weakCeil;

  const OutlineMaskConfig({
    required this.globalDark,
    required this.localMargin,
    required this.adaptiveCeil,
    required this.weakMargin,
    required this.weakCeil,
  });

  factory OutlineMaskConfig.fromJson(
      Map<String, dynamic> json, OutlineMaskConfig d) {
    return OutlineMaskConfig(
      globalDark: (json['globalDark'] as num?)?.toInt() ?? d.globalDark,
      localMargin: (json['localMargin'] as num?)?.toInt() ?? d.localMargin,
      adaptiveCeil: (json['adaptiveCeil'] as num?)?.toInt() ?? d.adaptiveCeil,
      weakMargin: (json['weakMargin'] as num?)?.toInt() ?? d.weakMargin,
      weakCeil: (json['weakCeil'] as num?)?.toInt() ?? d.weakCeil,
    );
  }
}

class LineBridgeConfig {
  final int maxGapMin;
  final int maxGapMax;
  final double faceCos;
  final int tangentSteps;

  const LineBridgeConfig({
    required this.maxGapMin,
    required this.maxGapMax,
    required this.faceCos,
    required this.tangentSteps,
  });

  factory LineBridgeConfig.fromJson(
      Map<String, dynamic> json, LineBridgeConfig d) {
    return LineBridgeConfig(
      maxGapMin: (json['maxGapMin'] as num?)?.toInt() ?? d.maxGapMin,
      maxGapMax: (json['maxGapMax'] as num?)?.toInt() ?? d.maxGapMax,
      faceCos: (json['faceCos'] as num?)?.toDouble() ?? d.faceCos,
      tangentSteps:
          (json['tangentSteps'] as num?)?.toInt() ?? d.tangentSteps,
    );
  }
}

class TrappedBallConfig {
  final int maxRMin;
  final int maxRMax;
  final int decayNumerator;
  final int decayDenominator;
  final int finalPassRadius;

  const TrappedBallConfig({
    required this.maxRMin,
    required this.maxRMax,
    required this.decayNumerator,
    required this.decayDenominator,
    required this.finalPassRadius,
  });

  factory TrappedBallConfig.fromJson(
      Map<String, dynamic> json, TrappedBallConfig d) {
    return TrappedBallConfig(
      maxRMin: (json['maxRMin'] as num?)?.toInt() ?? d.maxRMin,
      maxRMax: (json['maxRMax'] as num?)?.toInt() ?? d.maxRMax,
      decayNumerator:
          (json['decayNumerator'] as num?)?.toInt() ?? d.decayNumerator,
      decayDenominator:
          (json['decayDenominator'] as num?)?.toInt() ?? d.decayDenominator,
      finalPassRadius:
          (json['finalPassRadius'] as num?)?.toInt() ?? d.finalPassRadius,
    );
  }
}

class RegionFilterConfig {
  final int absoluteFloor;
  final int promoteFloor;
  final int rescueFloor;

  const RegionFilterConfig({
    required this.absoluteFloor,
    required this.promoteFloor,
    required this.rescueFloor,
  });

  factory RegionFilterConfig.fromJson(
      Map<String, dynamic> json, RegionFilterConfig d) {
    return RegionFilterConfig(
      absoluteFloor:
          (json['absoluteFloor'] as num?)?.toInt() ?? d.absoluteFloor,
      promoteFloor:
          (json['promoteFloor'] as num?)?.toInt() ?? d.promoteFloor,
      rescueFloor: (json['rescueFloor'] as num?)?.toInt() ?? d.rescueFloor,
    );
  }
}

class SnapRadiusConfig {
  final int web;
  final int flutter;

  const SnapRadiusConfig({required this.web, required this.flutter});

  factory SnapRadiusConfig.fromJson(
      Map<String, dynamic> json, SnapRadiusConfig d) {
    return SnapRadiusConfig(
      web: (json['web'] as num?)?.toInt() ?? d.web,
      flutter: (json['flutter'] as num?)?.toInt() ?? d.flutter,
    );
  }
}

class DetectionConfig {
  final OutlineMaskConfig outlineMask;
  final LineBridgeConfig lineBridge;
  final TrappedBallConfig trappedBall;
  final RegionFilterConfig regionFilter;
  final int bfsDarkThreshold;
  final double bfsFloodCapRatio;
  final int workerTimeoutMs;
  final SnapRadiusConfig snapRadius;

  const DetectionConfig({
    required this.outlineMask,
    required this.lineBridge,
    required this.trappedBall,
    required this.regionFilter,
    required this.bfsDarkThreshold,
    required this.bfsFloodCapRatio,
    required this.workerTimeoutMs,
    required this.snapRadius,
  });

  factory DetectionConfig.fromJson(
      Map<String, dynamic> json, DetectionConfig d) {
    final om = json['outlineMask'];
    final lb = json['lineBridge'];
    final tb = json['trappedBall'];
    final rf = json['regionFilter'];
    final sr = json['snapRadius'];
    return DetectionConfig(
      outlineMask: om is Map<String, dynamic>
          ? OutlineMaskConfig.fromJson(om, d.outlineMask)
          : d.outlineMask,
      lineBridge: lb is Map<String, dynamic>
          ? LineBridgeConfig.fromJson(lb, d.lineBridge)
          : d.lineBridge,
      trappedBall: tb is Map<String, dynamic>
          ? TrappedBallConfig.fromJson(tb, d.trappedBall)
          : d.trappedBall,
      regionFilter: rf is Map<String, dynamic>
          ? RegionFilterConfig.fromJson(rf, d.regionFilter)
          : d.regionFilter,
      bfsDarkThreshold:
          (json['bfsDarkThreshold'] as num?)?.toInt() ?? d.bfsDarkThreshold,
      bfsFloodCapRatio:
          (json['bfsFloodCapRatio'] as num?)?.toDouble() ?? d.bfsFloodCapRatio,
      workerTimeoutMs:
          (json['workerTimeoutMs'] as num?)?.toInt() ?? d.workerTimeoutMs,
      snapRadius: sr is Map<String, dynamic>
          ? SnapRadiusConfig.fromJson(sr, d.snapRadius)
          : d.snapRadius,
    );
  }
}

class CompletionConfig {
  final double freeCoverageRatio;
  final double freehandCoverThreshold;
  final double tinyRegionAutoRatio;

  const CompletionConfig({
    required this.freeCoverageRatio,
    required this.freehandCoverThreshold,
    required this.tinyRegionAutoRatio,
  });

  factory CompletionConfig.fromJson(
      Map<String, dynamic> json, CompletionConfig d) {
    return CompletionConfig(
      freeCoverageRatio:
          (json['freeCoverageRatio'] as num?)?.toDouble() ?? d.freeCoverageRatio,
      freehandCoverThreshold: (json['freehandCoverThreshold'] as num?)
              ?.toDouble() ??
          d.freehandCoverThreshold,
      tinyRegionAutoRatio:
          (json['tinyRegionAutoRatio'] as num?)?.toDouble() ??
              d.tinyRegionAutoRatio,
    );
  }
}

class CanvasConfig {
  final int maxDisplayEdgePx;
  final int undoStackMax;

  const CanvasConfig({
    required this.maxDisplayEdgePx,
    required this.undoStackMax,
  });

  factory CanvasConfig.fromJson(Map<String, dynamic> json, CanvasConfig d) {
    return CanvasConfig(
      maxDisplayEdgePx:
          (json['maxDisplayEdgePx'] as num?)?.toInt() ?? d.maxDisplayEdgePx,
      undoStackMax:
          (json['undoStackMax'] as num?)?.toInt() ?? d.undoStackMax,
    );
  }
}

class GenerationConfig {
  final int clientTimeoutMs;
  final int seedRange;
  final int maxImageBytes;

  const GenerationConfig({
    required this.clientTimeoutMs,
    required this.seedRange,
    required this.maxImageBytes,
  });

  factory GenerationConfig.fromJson(
      Map<String, dynamic> json, GenerationConfig d) {
    return GenerationConfig(
      clientTimeoutMs:
          (json['clientTimeoutMs'] as num?)?.toInt() ?? d.clientTimeoutMs,
      seedRange: (json['seedRange'] as num?)?.toInt() ?? d.seedRange,
      maxImageBytes:
          (json['maxImageBytes'] as num?)?.toInt() ?? d.maxImageBytes,
    );
  }
}

class DrawingConfig {
  final Map<String, DifficultyConfig> difficulties;
  final DetectionConfig detection;
  final CompletionConfig completion;
  final CanvasConfig canvas;
  final GenerationConfig generation;

  const DrawingConfig({
    required this.difficulties,
    required this.detection,
    required this.completion,
    required this.canvas,
    required this.generation,
  });

  static DrawingConfig get defaults => DrawingConfig(
        difficulties: const {
          'easy': DifficultyConfig(
              minArea: 2000,
              maxRegions: 10,
              colorCount: 6,
              resolution: 1024,
              skipFreeTiers: false),
          'medium': DifficultyConfig(
              minArea: 800,
              maxRegions: 18,
              colorCount: 8,
              resolution: 1024,
              skipFreeTiers: false),
          'hard': DifficultyConfig(
              minArea: 300,
              maxRegions: 30,
              colorCount: 12,
              resolution: 1024,
              skipFreeTiers: true),
          'extreme': DifficultyConfig(
              minArea: 80,
              maxRegions: 48,
              colorCount: 99,
              resolution: 1024,
              skipFreeTiers: true),
        },
        detection: const DetectionConfig(
          outlineMask: OutlineMaskConfig(
            globalDark: 100,
            localMargin: 22,
            adaptiveCeil: 165,
            weakMargin: 8,
            weakCeil: 212,
          ),
          lineBridge: LineBridgeConfig(
            maxGapMin: 4,
            maxGapMax: 20,
            faceCos: 0.5,
            tangentSteps: 4,
          ),
          trappedBall: TrappedBallConfig(
            maxRMin: 2,
            maxRMax: 7,
            decayNumerator: 2,
            decayDenominator: 3,
            finalPassRadius: 1,
          ),
          regionFilter: RegionFilterConfig(
            absoluteFloor: 30,
            promoteFloor: 50,
            rescueFloor: 5,
          ),
          bfsDarkThreshold: 80,
          bfsFloodCapRatio: 0.65,
          workerTimeoutMs: 90000,
          snapRadius: SnapRadiusConfig(web: 16, flutter: 12),
        ),
        completion: const CompletionConfig(
          freeCoverageRatio: 0.90,
          freehandCoverThreshold: 0.45,
          tinyRegionAutoRatio: 0.0005,
        ),
        canvas: const CanvasConfig(
          maxDisplayEdgePx: 2048,
          undoStackMax: 20,
        ),
        generation: const GenerationConfig(
          clientTimeoutMs: 80000,
          seedRange: 2000000000,
          maxImageBytes: 26214400,
        ),
      );

  factory DrawingConfig.fromJson(Map<String, dynamic> json) {
    final d = DrawingConfig.defaults;

    // difficulties
    Map<String, DifficultyConfig> difficulties = d.difficulties;
    final difsRaw = json['difficulties'];
    if (difsRaw is Map<String, dynamic>) {
      difficulties = Map<String, DifficultyConfig>.from(d.difficulties);
      for (final entry in difsRaw.entries) {
        final val = entry.value;
        if (val is Map<String, dynamic>) {
          final fallback =
              d.difficulties[entry.key] ?? d.difficulties['medium']!;
          difficulties[entry.key] =
              DifficultyConfig.fromJson(val, fallback);
        }
      }
    }

    // detection
    final detRaw = json['detection'];
    final detection = detRaw is Map<String, dynamic>
        ? DetectionConfig.fromJson(detRaw, d.detection)
        : d.detection;

    // completion
    final compRaw = json['completion'];
    final completion = compRaw is Map<String, dynamic>
        ? CompletionConfig.fromJson(compRaw, d.completion)
        : d.completion;

    // canvas
    final canvRaw = json['canvas'];
    final canvas = canvRaw is Map<String, dynamic>
        ? CanvasConfig.fromJson(canvRaw, d.canvas)
        : d.canvas;

    // generation
    final genRaw = json['generation'];
    final generation = genRaw is Map<String, dynamic>
        ? GenerationConfig.fromJson(genRaw, d.generation)
        : d.generation;

    return DrawingConfig(
      difficulties: difficulties,
      detection: detection,
      completion: completion,
      canvas: canvas,
      generation: generation,
    );
  }
}
